File: kalga-api/app/routers/stats.py
"""
Router pour les statistiques et analytics
"""
from fastapi import APIRouter, HTTPException, Query
from datetime import date, timedelta
from typing import Optional

from ..database.repositories import get_stats_repository, MerchantRepository

router = APIRouter(prefix="/api/stats", tags=["statistics"])

stats_repo = get_stats_repository()
merchant_repo = MerchantRepository()


@router.get("/{merchant_phone}/summary")
async def get_summary_stats(
    merchant_phone: str,
    days: int = Query(30, ge=1, le=365, description="Nombre de jours")
):
    """
    Récupère un résumé des statistiques sur une période.
    """
    merchant = await merchant_repo.get_by_phone(merchant_phone)
    if not merchant:
        raise HTTPException(status_code=404, detail="Marchand non trouvé")

    summary = await stats_repo.get_summary_stats(merchant["id"], days)
    conversion = await stats_repo.get_conversion_rate(merchant["id"], days)

    return {
        "merchant_id": merchant["id"],
        "merchant_phone": merchant_phone,
        **summary,
        **conversion
    }


@router.get("/{merchant_phone}/daily")
async def get_daily_stats(
    merchant_phone: str,
    start_date: Optional[str] = Query(None, description="Date de début (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="Date de fin (YYYY-MM-DD)"),
    days: int = Query(7, ge=1, le=90, description="Nombre de jours si pas de dates")
):
    """
    Récupère les statistiques quotidiennes pour les graphiques.
    """
    merchant = await merchant_repo.get_by_phone(merchant_phone)
    if not merchant:
        raise HTTPException(status_code=404, detail="Marchand non trouvé")

    # Parser les dates ou utiliser les jours par défaut
    if start_date and end_date:
        try:
            start = date.fromisoformat(start_date)
            end = date.fromisoformat(end_date)
        except ValueError:
            raise HTTPException(status_code=400, detail="Format de date invalide")
    else:
        end = date.today()
        start = end - timedelta(days=days)

    stats = await stats_repo.get_stats_range(merchant["id"], start, end)

    # Remplir les jours manquants avec des zéros
    date_range = []
    current = start
    while current <= end:
        date_range.append(current.isoformat())
        current += timedelta(days=1)

    # Créer un dictionnaire des stats existantes
    stats_dict = {s["date"]: s for s in stats}

    # Remplir les données
    result = []
    for d in date_range:
        if d in stats_dict:
            result.append(stats_dict[d])
        else:
            result.append({
                "date": d,
                "conversations_count": 0,
                "messages_count": 0,
                "sales_count": 0,
                "revenue": 0,
                "unique_clients": 0
            })

    return {
        "merchant_id": merchant["id"],
        "start_date": start.isoformat(),
        "end_date": end.isoformat(),
        "data": result
    }


@router.get("/{merchant_phone}/top-products")
async def get_top_products(
    merchant_phone: str,
    limit: int = Query(5, ge=1, le=20, description="Nombre de produits")
):
    """
    Récupère les produits les plus populaires.
    """
    merchant = await merchant_repo.get_by_phone(merchant_phone)
    if not merchant:
        raise HTTPException(status_code=404, detail="Marchand non trouvé")

    products = await stats_repo.get_top_products(merchant["id"], limit)

    return {
        "merchant_id": merchant["id"],
        "products": products
    }


@router.get("/{merchant_phone}/hourly-activity")
async def get_hourly_activity(
    merchant_phone: str,
    days: int = Query(7, ge=1, le=30, description="Nombre de jours")
):
    """
    Récupère l'activité par heure de la journée.
    """
    merchant = await merchant_repo.get_by_phone(merchant_phone)
    if not merchant:
        raise HTTPException(status_code=404, detail="Marchand non trouvé")

    activity = await stats_repo.get_hourly_activity(merchant["id"], days)

    # S'assurer que toutes les heures sont présentes
    hour_data = {h: 0 for h in range(24)}
    for item in activity:
        hour_data[item["hour"]] = item["count"]

    return {
        "merchant_id": merchant["id"],
        "period_days": days,
        "data": [{"hour": h, "count": c} for h, c in hour_data.items()]
    }


@router.get("/{merchant_phone}/realtime")
async def get_realtime_stats(merchant_phone: str):
    """
    Récupère les statistiques en temps réel (aujourd'hui).
    """
    merchant = await merchant_repo.get_by_phone(merchant_phone)
    if not merchant:
        raise HTTPException(status_code=404, detail="Marchand non trouvé")

    today_stats = await stats_repo.get_or_create_daily_stats(merchant["id"])

    return {
        "merchant_id": merchant["id"],
        "date": date.today().isoformat(),
        "conversations_today": today_stats.get("conversations_count", 0),
        "messages_today": today_stats.get("messages_count", 0),
        "sales_today": today_stats.get("sales_count", 0),
        "revenue_today": today_stats.get("revenue", 0)
    }

File: kalga-web/convex/internal/stats.ts
import { query, mutation } from "../_generated/server";
import { v } from "convex/values";
import type { Doc } from "../_generated/dataModel";
import { assertInternalKey } from "./guard";

// Visibilité : voir note dans `inventory.ts` — fonctions publiques gardées par
// l'argument secret `internalKey` (server-to-server Python sans deploy key).

/**
 * Fonctions internes statistiques / analytics (appelées par le backend Python).
 * Parité `stats_repo` (kalga-api). Convex n'a PAS d'agrégat SQL : SUM/AVG/GROUP
 * BY sont recalculés en JS sur les docs collectés.
 *
 * Dates : SQLite stockait `daily_stats.date` en ISO 'YYYY-MM-DD'. On garde le
 * même contrat — Python passe les bornes en ISO, on compare lexicographiquement.
 */

/** Champs numériques d'une ligne `dailyStats` avec défaut 0. */
function statRow(d: Doc<"dailyStats">) {
  return {
    id: d._id,
    merchant_id: d.merchantId,
    date: d.date,
    conversations_count: d.conversationsCount ?? 0,
    messages_count: d.messagesCount ?? 0,
    sales_count: d.salesCount ?? 0,
    revenue: d.revenue ?? 0,
    unique_clients: d.uniqueClients ?? 0,
    avg_response_time: d.avgResponseTime ?? null,
  };
}

/**
 * Récupère (ou crée) la ligne de stats du jour (parité `get_or_create_daily_stats`).
 * `date` = ISO 'YYYY-MM-DD' fourni par Python (today côté serveur Python).
 */
export const getOrCreateDailyStats = mutation({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    date: v.string(),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const existing = await ctx.db
      .query("dailyStats")
      .withIndex("by_merchant_date", (q) =>
        q.eq("merchantId", args.merchantId).eq("date", args.date),
      )
      .unique();
    if (existing) return statRow(existing);

    const id = await ctx.db.insert("dailyStats", {
      merchantId: args.merchantId,
      date: args.date,
    });
    const created = await ctx.db.get(id);
    return statRow(created!);
  },
});

/** Upsert atomique d'un delta sur la ligne du jour (parité increment_x / record_sale). */
async function bumpDay(
  ctx: { db: any },
  merchantId: Doc<"merchants">["_id"],
  date: string,
  patch: Partial<{
    conversationsDelta: number;
    messagesDelta: number;
    salesDelta: number;
    revenueDelta: number;
    uniqueClients: number;
  }>,
) {
  const existing = await ctx.db
    .query("dailyStats")
    .withIndex("by_merchant_date", (q: any) =>
      q.eq("merchantId", merchantId).eq("date", date),
    )
    .unique();
  if (existing) {
    const upd: Record<string, number> = {};
    if (patch.conversationsDelta)
      upd.conversationsCount =
        (existing.conversationsCount ?? 0) + patch.conversationsDelta;
    if (patch.messagesDelta)
      upd.messagesCount = (existing.messagesCount ?? 0) + patch.messagesDelta;
    if (patch.salesDelta)
      upd.salesCount = (existing.salesCount ?? 0) + patch.salesDelta;
    if (patch.revenueDelta)
      upd.revenue = (existing.revenue ?? 0) + patch.revenueDelta;
    if (patch.uniqueClients !== undefined)
      upd.uniqueClients = patch.uniqueClients;
    await ctx.db.patch(existing._id, upd);
    return;
  }
  await ctx.db.insert("dailyStats", {
    merchantId,
    date,
    conversationsCount: patch.conversationsDelta ?? 0,
    messagesCount: patch.messagesDelta ?? 0,
    salesCount: patch.salesDelta ?? 0,
    revenue: patch.revenueDelta ?? 0,
    uniqueClients: patch.uniqueClients ?? 0,
  });
}

/** Incrémente conversations / messages / ventes du jour (parité increment_x / record_sale). */
export const bumpDailyStats = mutation({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    date: v.string(),
    conversationsDelta: v.optional(v.number()),
    messagesDelta: v.optional(v.number()),
    salesDelta: v.optional(v.number()),
    revenueDelta: v.optional(v.number()),
    uniqueClients: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    await bumpDay(ctx, args.merchantId, args.date, {
      conversationsDelta: args.conversationsDelta,
      messagesDelta: args.messagesDelta,
      salesDelta: args.salesDelta,
      revenueDelta: args.revenueDelta,
      uniqueClients: args.uniqueClients,
    });
    return { ok: true };
  },
});

/** Lignes de stats sur une plage de dates ISO (parité `get_stats_range`). */
export const getStatsRange = query({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    startDate: v.string(),
    endDate: v.string(),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const rows = await ctx.db
      .query("dailyStats")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();
    return rows
      .filter((r) => r.date >= args.startDate && r.date <= args.endDate)
      .sort((a, b) => a.date.localeCompare(b.date))
      .map(statRow);
  },
});

/** Top produits par conversations/ventes (parité `get_top_products`). */
export const getTopProducts = query({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const products = await ctx.db
      .query("products")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();
    const convs = await ctx.db
      .query("conversations")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();

    const byProduct = new Map<
      string,
      { conversation_count: number; sales_count: number; revenue: number }
    >();
    for (const c of convs) {
      const key = c.productId as string;
      const agg = byProduct.get(key) ?? {
        conversation_count: 0,
        sales_count: 0,
        revenue: 0,
      };
      agg.conversation_count += 1;
      if (c.status === "completed") {
        agg.sales_count += 1;
        agg.revenue += c.currentOffer ?? 0;
      }
      byProduct.set(key, agg);
    }

    const rows = products.map((p) => {
      const agg = byProduct.get(p._id as string) ?? {
        conversation_count: 0,
        sales_count: 0,
        revenue: 0,
      };
      return { id: p._id, name: p.name, code: p.code, ...agg };
    });
    rows.sort(
      (a, b) =>
        b.sales_count - a.sales_count ||
        b.conversation_count - a.conversation_count,
    );
    return rows.slice(0, args.limit ?? 5);
  },
});

/** Taux de conversion sur N jours (parité `get_conversion_rate`). */
export const getConversionRate = query({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    startMs: v.number(), // borne basse epoch ms (today - days)
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const convs = await ctx.db
      .query("conversations")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();
    const inRange = convs.filter((c) => c._creationTime >= args.startMs);
    const total = inRange.length;
    const completed = inRange.filter((c) => c.status === "completed").length;
    const pending = inRange.filter(
      (c) =>
        c.status === "pending_delivery" || c.status === "pending_pickup",
    ).length;
    const rate = total > 0 ? Math.round((completed / total) * 1000) / 10 : 0;
    return {
      total_conversations: total,
      completed_sales: completed,
      pending_sales: pending,
      conversion_rate: rate,
    };
  },
});

/**
 * Activité par heure de la journée sur N jours (parité `get_hourly_activity`).
 * SQLite groupait `strftime('%H', messages.created_at)`. Convex : on collecte
 * les conversations du marchand, puis les messages de chaque conversation sur
 * la fenêtre, et on bucketise par heure locale du `_creationTime`.
 *
 * `tzOffsetMinutes` = décalage local (ex: Abidjan = 0). Python le passe pour
 * que l'heure corresponde au fuseau marchand (SQLite stockait l'heure locale).
 */
export const getHourlyActivity = query({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    startMs: v.number(),
    tzOffsetMinutes: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const convs = await ctx.db
      .query("conversations")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();
    const offsetMs = (args.tzOffsetMinutes ?? 0) * 60 * 1000;
    const buckets = new Map<number, number>();
    for (const c of convs) {
      const msgs = await ctx.db
        .query("messages")
        .withIndex("by_conversation", (q) => q.eq("conversationId", c._id))
        .collect();
      for (const m of msgs) {
        if (m._creationTime < args.startMs) continue;
        const hour = new Date(m._creationTime + offsetMs).getUTCHours();
        buckets.set(hour, (buckets.get(hour) ?? 0) + 1);
      }
    }
    return [...buckets.entries()]
      .map(([hour, count]) => ({ hour, count }))
      .sort((a, b) => a.hour - b.hour);
  },
});

/** Enregistre un événement analytics (parité `log_event`). `data` = JSON string. */
export const logEvent = mutation({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    eventType: v.string(),
    productId: v.optional(v.id("products")),
    conversationId: v.optional(v.id("conversations")),
    clientPhone: v.optional(v.string()),
    data: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    const id = await ctx.db.insert("analyticsEvents", {
      merchantId: args.merchantId,
      eventType: args.eventType,
      productId: args.productId,
      conversationId: args.conversationId,
      clientPhone: args.clientPhone,
      data: args.data,
    });
    return { eventId: id };
  },
});

/** Événements analytics récents (parité `get_recent_events`). */
export const getRecentEvents = query({
  args: {
    internalKey: v.string(),
    merchantId: v.id("merchants"),
    eventType: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    assertInternalKey(args.internalKey);
    let rows = await ctx.db
      .query("analyticsEvents")
      .withIndex("by_merchant", (q) => q.eq("merchantId", args.merchantId))
      .collect();
    if (args.eventType)
      rows = rows.filter((r) => r.eventType === args.eventType);
    rows.sort((a, b) => b._creationTime - a._creationTime);
    return rows.slice(0, args.limit ?? 50);
  },
});
